File: G_Projects/I_Vending_Machine/VendingMachineStates/ConcreteStates/DispenseState.java
package G_Projects.I_Vending_Machine.VendingMachineStates.ConcreteStates;

import G_Projects.I_Vending_Machine.CommonEnums.Coin;
import G_Projects.I_Vending_Machine.CommonExceptions.InvalidShelfCodeException;
import G_Projects.I_Vending_Machine.CommonExceptions.ItemSoldOutException;
import G_Projects.I_Vending_Machine.VendingMachineStates.VendingMachineContext;
import G_Projects.I_Vending_Machine.VendingMachineStates.VendingMachineState;

public class DispenseState implements VendingMachineState{

    @Override
    public String getStateName(){
        return "Dispense State";
    }

    @Override
    public void insertCoin(VendingMachineContext context, Coin coin){
        throw new IllegalStateException("Cannot insert coin while dispensing item");
    }

    @Override
    public void selectItem(VendingMachineContext context, int itemCode){
        throw new IllegalStateException("Cannot select item while dispensing another item");
    }

    @Override
    public void dispenseItem(VendingMachineContext context){

        int code = context.getSelectedItemCode();

        try {
            context.getInventory().removeItem(code);
        } catch (InvalidShelfCodeException e) {
            throw new IllegalStateException("Invalid shelf code during dispense: " + code + e);
        } catch (ItemSoldOutException e) {
            context.setState(new OutOfStockState());
            throw new IllegalStateException("Item sold out during dispense: " + code + e);
        }

        context.clearBalance();
        context.resetSelection();

        context.setState(new IdleState());
    }
}

File: Python/A_Project/E_Vending_Machine/VendingMachineStates/ConcreteStates/dispense_state.py
from CommonExceptions.invalid_shelf_code_exception import InvalidShelfCodeException
from CommonExceptions.item_sold_out_exception import ItemSoldOutException
from VendingMachineStates.ConcreteStates.out_of_stock_state import OutOfStockState
from VendingMachineStates.vending_machine_state import VendingMachineState


class DispenseState(VendingMachineState):
    def get_state_name(self) -> str:
        return "Dispense State"

    def insert_coin(self, context, coin):
        raise RuntimeError("Cannot insert coin while dispensing item")

    def select_item(self, context, item_code: int):
        raise RuntimeError("Cannot select item while dispensing another item")

    def dispense_item(self, context):
        # To handle circular import issue, we import IdleState here instead of at the top
        from VendingMachineStates.ConcreteStates.idle_state import IdleState

        code = context.get_selected_item_code()
        try:
            context.get_inventory().remove_item(code)
        except InvalidShelfCodeException as exc:
            raise RuntimeError(f"Invalid shelf code during dispense: {code}") from exc
        except ItemSoldOutException as exc:
            context.set_state(OutOfStockState())
            raise RuntimeError(f"Item sold out during dispense: {code}") from exc

        context.clear_balance()
        context.reset_selection()
        context.set_state(IdleState())
